nixos/modules/services/networking/syncthing.nix

```diff
@@ -615,7 +615,10 @@ in {
           ];
         };
       };
-      syncthing-init = mkIf (cfg.settings != {}) {
+      syncthing-init = mkIf (
+        # we filter all the null values out to easier compare the two sets
+        (lib.filterAttrsRecursive (_: v: v != null) cfg.settings) != { devices = {}; folders = {}; options = {}; }
```

Comment:
This might fail, if in the future we'll introduce an inner option inside `settings.options` like the following:

```nix
service.syncthing.settings.options = {
    innerSyncthingOption = {
        # Default options:
        key1 = null;
        key2 = null;
    };
};
```

Ideally, I think we should create an intermediate variable that will hold all the settings, with `{}` values also completely filtered, and also inner attributes such that e.g the following settings will be transformed into simply `null`:

```nix
settings = {
    b = {
        c1 = {
            d = null;
        };
        c2 = {
            d1 = null;
            d2 = null;
        };
    };
};
```

This way, when generating the `syncthing-init.service` script, if we'll read a completely filtered version of `cfg.settings`, we will avoid relying on the API not erroring when JSON values are `null`.

Reply:
maybe we should just filter some more? how about:

```
filterAttrsRecursive (_: v: v != null && v != {} ) cfg.settings
```

Reply:
I pushed a small change to filter settings recursively (also for the json payload)